### PR TITLE
[Collaborative optimization] Pruning-Clustering-preserving Quantization Aware Training

### PR DESCRIPTION
**Pull request by Ruomei**
This PR adds support for Pruning-Clustering-preserving Quantization Aware Training (PCQAT). Aiming at preserving the sparsity and unique weights in the output optimized model, fixed pruning masks and the stochastic updates of clustering training variables are enabled during quantization-aware training.

User API:
```
  preserve_sparsity = True
  quant_aware_annotate_model = quantize.quantize_annotate_model(pruned_clustered_model)
  pcqat_model = quantize.quantize_apply(
      quant_aware_annotate_model,
      scheme=default_8bit_cluster_preserve_quantize_scheme
      .Default8BitClusterPreserveQuantizeScheme(preserve_sparsity))
```

Main changes:
- cluster_preserve_integration_test: edge cases (e.g. passing non-pruned models, passing models with uniform weights, etc.), checks for the updates of trainable variables between epochs
- mnist_prune_cluster_preserve_qat_test: minimal e2e mnist example to show benefits of PCQAT with the most common two types of configurations
- cluster_preserve_quantize_registry: main implementation of PCQAT algo.

Initial results (pruning sparsity: 50%, number of clusters: 8 (DS-CNN-L), 16 (Mobilenet_v1)):
| Model | Items | Baseline | Pruned Model | QATed | Pruned_Clustered Model | PCQATed Model | 
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
| DS-CNN-L | FP32 Top1 Accuracy | 95.06% | 94.07% | (Fake INT8) 94.85% | **93.76%** | **(Fake INT8) 94.28%** |
|                        | INT8 full integer quantization | 94.35% | 93.80% | 94.82% | **93.21%** | **94.06%** |
|                        | INT8 .tflite gzip compression (bytes) | 506400 -> 425006 | 506400 -> 317937 | 507296 -> 424368 | 506400 -> 205333 | 507296 -> 201744 |
| Mobilenet_v1 (ImageNet) | FP32 Top1 Accuracy | 70.98% | 70.49% | (Fake INT8) 70.88% | **67.64%** | **(Fake INT8) 67.80%** |
|                        | INT8 full integer quantization | 70.37% | 69.85% | 70.87% | **66.89%** | **68.63%** |
|                        | INT8 .tflite gzip compression (bytes) | 4665552 -> 3886236 | 4665552 -> 2909148 | 4569416 -> 3808781| 4665552 -> 2013010 | 4569472 -> 1943957 |

### CODE REVIEW

File: tensorflow_model_optimization/python/core/clustering/keras/cluster_wrapper.py [hunk 1]
```diff
@@ -215,6 +215,10 @@ def update_clustered_weights_associations(self):
               pulling_indices, original_weight))
 
       if self.preserve_sparsity:
+        # Re-discover the sparsity masks to avoid zero-drifting
```
Comment (akarmi) [inline tensorflow_model_optimization/python/core/clustering/keras/cluster_wrapper.py:218]: This change is subject to #702 - to merge this one cleanly, please remove it from this PR.

**Review by akarmi — CHANGES_REQUESTED** [review]
Thanks, a minor request below.

File: tensorflow_model_optimization/python/core/clustering/keras/cluster_wrapper.py (hunk 1 above)
Comment (Ruomei) [reply tensorflow_model_optimization/python/core/clustering/keras/cluster_wrapper.py:218]: Will do. Thanks.

**Review by akarmi — APPROVED** [review]
Thank you.

**Comment by Ruomei** [comment]
Hi @daverim, recently we have removed the dependency of this PR, so it is ready for review anytime. Could you please take a look when you have time?
Thanks!

File: tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry.py [hunk 2]
```diff
@@ -297,75 +266,89 @@ def _build_clusters(self, name, layer):
...
+      ori_weights_tf = layer.add_weight(
+          'ori_weights_vars_tf',
+          shape=weights.shape,
+          initializer=tf.keras.initializers.Constant(
+              value=K.batch_get_value([weights])[0]),
+          dtype=weights.dtype,
+          trainable=True)
+
+      # Get clustering implementation according to layer type
+      clustering_impl_cls = clustering_registry.ClusteringLookupRegistry().\
```
Comment (daverim) [inline tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry.py:307]: Our linting tools will flag line continuation, prefer using parentheses

File: tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry.py [hunk 3]
```diff
@@ -242,13 +205,16 @@ def apply_cluster_preserve_quantize_config(self, layer, quantize_config):
...
   """Default 8 bit ClusterPreserveQuantizeRegistry."""
+  def __init__(self, preserve_sparsity):
+    super(Default8bitClusterPreserveQuantizeRegistry, self).__init__(
+        preserve_sparsity)
+    self.preserve_sparsity = preserve_sparsity
 
   def get_quantize_config(self, layer):
-    """Returns the quantization config with addon cluster preserve weight_quantizer for the given layer.
-
+    """Returns the quantization config with addon cluster preserve
```
Comment (daverim) [inline tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry.py:214]: Should be one line

File: tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/default_8bit_cluster_preserve_quantize_scheme.py [hunk 4]
```diff
@@ -23,7 +23,14 @@
...
+    clusters and sparsity if the special flag is set.
+    Args:
+      preserve_sparsity: the flag to enable prune-cluster-preserving QAT.
+    """
+    self.preserve_sparsity = preserve_sparsity
 
   def get_quantize_registry(self):
     return cluster_preserve_quantize_registry.\
-        Default8bitClusterPreserveQuantizeRegistry()
+      Default8bitClusterPreserveQuantizeRegistry(self.preserve_sparsity)
```
Comment (daverim) [inline tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/default_8bit_cluster_preserve_quantize_scheme.py:36]: Should be indented 2 spaces

**Review by daverim — COMMENTED** [review]
Some small linting issues.

File: tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry.py [hunk 5]
```diff
@@ -297,75 +266,89 @@ def _build_clusters(self, name, layer):
...
+      centroids_mask = 1.0 - tf.one_hot(zero_idx, num_centroids)
+      result = {'sparsity_mask': sparsity_mask}
+
+    # Prepare clustering variables for the Keras graph when clusters
+    # exist, assuming we do not use number_of_clusters larger than 1024
+    if num_centroids > 1024:
+      return result
+    else:
+      clst_centroids_tf = layer.add_weight(
+          'cluster_centroids_tf',
```
[the remark this answers is not in the document]
Comment (Ruomei) [reply tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry.py]: Sure, can do.
In the updated patch, I only modified keys added by PCQAT and left the ones in QAT untouched.
The definitions of constants are at the beginning of the `cluster_preserve_quantize_registry.py` like these:
```
CLUSTER_CENTROIDS = 'cluster_centroids_tf'
PULLING_INDICES = 'pulling_indices_tf'
ORIGINAL_WEIGHTS = 'ori_weights_vars_tf'
WEIGHT_NAME = 'weight_name'
CLUSTERING_IMPL = 'clst_impl'
CENTROIDS_MASK = 'centroids_mask'
SPARSITY_MASK = 'sparsity_mask'
```
Please let me know what you think?

File: tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry.py [hunk 6]
```diff
@@ -374,35 +357,50 @@ def build(self, tensor_shape, name, layer):
 
   def __call__(self, inputs, training, weights, **kwargs):
     """Apply cluster preserved quantization to the input tensor.
-
```
[the remark this answers is not in the document]
Comment (Ruomei) [reply tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry.py]: Thanks, done.

File: tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry.py (hunk 2 above)
Comment (Ruomei) [reply tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry.py]: Thanks, done.

File: tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry.py (hunk 3 above)
Comment (Ruomei) [reply tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/cluster_preserve_quantize_registry.py]: Thanks, done.

File: tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/default_8bit_cluster_preserve_quantize_scheme.py (hunk 4 above)
Comment (Ruomei) [reply tensorflow_model_optimization/python/core/quantization/keras/collaborative_optimizations/cluster_preserve/default_8bit_cluster_preserve_quantize_scheme.py]: Thanks, done. Also removed the `\`.

**Review by daverim — APPROVED** [review]

**Comment by Xhark** [comment]
Hi, Just curious why Pruned_Clustered Model - Mobilenet_v1 (ImageNet) - INT8 .tflite gzip compression (bytes) is so small? PCQAT model is larger than PC model? or missed a digit for this case?

**Comment by wwwind** [comment]
Hi @Xhark Yes, this is the mistake, last digit is missing. The compression ratio was around 2.3 in our experiments. 
Thanks for noticing.

**Comment by Ruomei** [comment]
Hi, @daverim and @Xhark, could you please also let us know if there is anything we can do to help with the failed internal checks shown in this PR?
@akarmi @wwwind for visibility
Thanks all!

**Comment by daverim** [comment]
Merging was blocked by build file strict dependencies -- resubmitting with
it fixed myself now, should be merged today.

**Comment by Ruomei** [comment]
> Merging was blocked by build file strict dependencies -- resubmitting with it fixed myself now, should be merged today.

Brill, thanks a lot, David.
